File: DynamicPrograming/go/min_change.go
package dynamicprogramming

// MinChange returns the minimum coins required or -1 if impossible.
func MinChange(amount int, coins []int) int {
	memo := make(map[int]int)

	var dfs func(int) int
	dfs = func(remaining int) int {
		if remaining < 0 {
			return maxInt
		}
		if remaining == 0 {
			return 0
		}
		if val, ok := memo[remaining]; ok {
			return val
		}
		best := maxInt
		for _, coin := range coins {
			candidate := 1 + dfs(remaining-coin)
			if candidate < best {
				best = candidate
			}
		}
		memo[remaining] = best
		return best
	}

	answer := dfs(amount)
	if answer >= maxInt {
		return -1
	}
	return answer
}

File: DynamicPrograming/ts/minChange.ts
export function minChange(amount: number, coins: number[]): number {
  const memo = new Map<number, number>();

  const dfs = (remaining: number): number => {
    if (remaining < 0) return Infinity;
    if (remaining === 0) return 0;
    if (memo.has(remaining)) return memo.get(remaining)!;

    let best = Infinity;
    for (const coin of coins) {
      const attempt = 1 + dfs(remaining - coin);
      best = Math.min(best, attempt);
    }

    memo.set(remaining, best);
    return best;
  };

  const answer = dfs(amount);
  return Number.isFinite(answer) ? answer : -1;
}
